src/bandersnatch/mirror.py

```diff
@@ -241,42 +291,73 @@ async def determine_packages_to_sync(self) -> None:
-        logger.debug(f"Package syncer {idx} started for duty")
-        while True:
-            try:
-                package = self.package_queue.get_nowait()
-            except asyncio.QueueEmpty:
-                logger.debug(f"Package syncer {idx} emptied queue")
-                break
+    async def process_package(self, package: Package) -> None:
+        # Don't save anything if our metadata filters all fail.
+        if not package.filter_metadata(self.filters.filter_metadata_plugins()):
+            return None
+
+        # save the metadata before filtering releases
+        if self.json_save:
+            loop = asyncio.get_event_loop()
+            json_saved = await loop.run_in_executor(
+                None, self.save_json_metadata, package.metadata, package.name
+            )
+            assert json_saved
 
-            await package.sync(self.filters)
+        package.filter_all_releases_files(self.filters.filter_release_file_plugins())
+        package.filter_all_releases(self.filters.filter_release_plugins())
 
-            # Cleanup non normalized name directory
-            await self.cleanup_non_pep_503_paths(package)
+        await self.sync_release_files(package)
+        self.sync_simple_page(package)
+        # XMLRPC PyPI Endpoint stores raw_name so we need to provide it
+        self.record_finished_package(package.raw_name)
 
-    async def sync_packages(self) -> None:
-        self.package_queue: asyncio.Queue = asyncio.Queue()
-        # Sorting the packages alphabetically makes it more predicatable:
-        # easier to debug and easier to follow in the logs.
-        for name in sorted(self.packages_to_sync):
-            serial = self.packages_to_sync[name]
-            await self.package_queue.put(Package(name, serial, self))
-
-        sync_coros: List[Awaitable] = [
-            self.package_syncer(idx) for idx in range(self.workers)
-        ]
-        try:
-            await asyncio.gather(*sync_coros)
-        except KeyboardInterrupt:
+        # Cleanup old legacy non PEP 503 Directories created for the Simple API
+        await self.cleanup_non_pep_503_paths(package)
+
+    def finalize_sync(self) -> None:
+        self.sync_index_page()
+        if self.need_wrapup:
+            self.wrapup_successful_sync()
+        return None
+
+    def on_error(self, exception: BaseException, **kwargs: Dict) -> None:
+        self.errors = True
+        if isinstance(exception, KeyboardInterrupt):
             # Setting self.errors to True to ensure we don't save Serial
             # and thus save to disk that we've had a successful sync
-            self.errors = True
             logger.info(
                 "Cancelling, all downloads are forcibly stopped, data may be "
                 + "corrupted. Serial will not be saved to disk. "
                 + "Next sync will start from previous serial"
             )
+        elif isinstance(exception, TypeError) or isinstance(exception, ValueError):
```

Comment:
Since we've separated this from it's context, it would probably be a good idea to have a comment here describing where these errors are coming from.